homeassistant/components/legrand_rflc/__init__.py

```diff
@@ -0,0 +1,81 @@
+"""The Legrand RFLC integration.
+
+https://www.legrand.us/solutions/smart-lighting/radio-frequency-lighting-controls
+"""
+
+import asyncio
+from collections.abc import Mapping
+from typing import Final
+
+import lc7001.aio
+
+from homeassistant import config_entries
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_AUTHENTICATION, CONF_HOST, CONF_PORT
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+PLATFORMS: Final = ["light"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up Legrand RFLC integration."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Set up Legrand LC7001 from a config entry."""
+    entry_id = entry.entry_id
+    data = entry.data
+    host = data[CONF_HOST]
+    kwargs = {}
+    if CONF_AUTHENTICATION in data:
+        kwargs["key"] = bytes.fromhex(data[CONF_AUTHENTICATION])
+    if CONF_PORT in data:  # for testing only (server emulation on localhost)
+        kwargs["port"] = data[CONF_PORT]
+    hass.data[DOMAIN][entry_id] = hub = lc7001.aio.Hub(host, **kwargs)
+
+    async def setup_platforms() -> None:
+        hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    async def _reauth() -> None:
+        await hass.config_entries.async_unload(entry_id)
+        await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={
+                "source": SOURCE_REAUTH,
+                "entry": entry,
+                "unique_id": entry.unique_id,
+            },
+            data=data,
+        )
+
+    async def reauth() -> None:
+        hass.async_create_task(_reauth())
+
+    async def reload(message: Mapping) -> None:
+        hass.async_create_task(hass.config_entries.async_reload(entry_id))
+
+    hub.once(hub.EVENT_AUTHENTICATED, setup_platforms)
+    hub.once(hub.EVENT_UNAUTHENTICATED, reauth)
+    hub.once(hub.EVENT_ZONE_ADDED, reload)
+    hub.once(hub.EVENT_ZONE_DELETED, reload)
+
+    asyncio.create_task(hub.loop())  # not hass.async_create_task
```

Comment:
I did call hass.async_create_task instead but **something** broke.
I found that if I called asyncio.create_task myself that things worked again.
Hence the comment.

I know that it breaks my test_light:EventAuditor which stimulates the emulated server and expects certain ordered events to occur.
The tests are not complete until **all** its expectations are met.
'hass.async_block_till_done()' is called to wait just long enough.
Unfortunately, this will never complete if my integration is a Home Assistant task.

My asyncio.create_task uses the asyncio.get_running_loop just like hass.async_create_task does so all callbacks from this task are made in the same event loop.
I don't know that this is a dangerous assumption to make.

Reply:
use hass.async_create_task now
https://github.com/home-assistant/core/pull/53430/commits/b1dbbadf183d472171b950ef35d28bd50679e681